### PR TITLE
make similar(A, [T,] shape...) for structured A yield a sparse rather than dense array

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request makes `similar(A<:Union{Diagonal,Bidiagonal,Tridiagonal,SymTridiagonal}, [neweltype,] shape...)` yield a sparse rather than dense array, consistent with e.g. `broadcast`. For example, this pull request makes `similar(Diagonal(rand(4)), Float32, (3, 3))` yield a `SparseMatrixCSC{Float32}` rather than the present `Matrix{Float32}`.

This pull request also fixes a minor bug where `lufact(A<:Tridiagonal{T})` where `!(T<:AbstractFloat)` would hit `lufact!(B, ...)` for non-`Tridiagonal` rather than `Tridiagonal` `B` downstream.

Ref. https://github.com/JuliaLang/LinearAlgebra.jl/issues/271. Best!

### CODE REVIEW

File: base/linalg/lu.jl [hunk 1]
```diff
@@ -416,6 +416,14 @@ function lufact!(A::Tridiagonal{T,V}, pivot::Union{Val{false}, Val{true}} = Val(
     LU{T,Tridiagonal{T,V}}(B, ipiv, convert(BlasInt, info))
 end
 
+# Absent the following methods, if A::Tridiagonal{T} where !(T<:AbstractFloat), then the
+# method lufact(A) hits sends copy!(similar(A, afloattype, size(A)), A) (which is not a
+# Tridagonal given the (unnecessary?) shape argument to similar) to lufact! rather than
```
[the remark this answers is not in the document]
Comment (Sacha0) [reply base/linalg/lu.jl]: Nixing the shape argument sounds good. That removing the shape argument might break other consumers of the general `lufact` method motivated the present conservative solution. Let's see whether CI complains :). Best!

**Comment by andreasnoack** [comment]
For some reason, this triggers a test error in `special`.

**Comment by Sacha0** [comment]
> For some reason, this triggers a test error in special.

Should be fixed. Absent objections or requests for time and assuming CI clears, I plan to merge these changes this evening. (If someone beats me to it, not squashing the commits might be best for future bug sleuthing.) Best!

**Comment by Sacha0** [comment]
Thanks all!

**Pull request merged**